contracts/exchangeIssuance/ExchangeIssuanceZeroEx.sol

```diff
@@ -0,0 +1,561 @@
+            }
+            else{
+                _tokens[i].safeTransfer(_to, _tokens[i].balanceOf(address(this)));
+            }
+        }
+    }
+
+    receive() external payable {
+        // required for weth.withdraw() to work properly
+        require(msg.sender == WETH, "ExchangeIssuance: Direct deposits not allowed");
+    }
+
+    /* ============ Public Functions ============ */
+
+    /**
+     * Whitelists an issuance module
+     *
+     * @param _issuanceModule    Struct containing data on issuance module to add
+     */
+    function addIssuanceModule(address _issuanceModule, bool _isDebtIssuanceModule) public onlyOwner {
+        _addIssuanceModule(_issuanceModule, _isDebtIssuanceModule);
+    }
+
+    /**
+     * Removes an issuance module from the whitelist
+     *
+     * @param _issuanceModuleAddress    Address of issuance module to remove
+     */
+    function removeIssuanceModule(address _issuanceModuleAddress) public onlyOwner {
+        _removeIssuanceModule(_issuanceModuleAddress);
+    }
+
+    /**
+     * Change the _swapTarget
+     *
+     * @param _swapTarget    Address of the swap target contract. (Usually ZeroEx ExchangeProxy)
+     */
+    function setSwapTarget(address _swapTarget) public onlyOwner {
+        swapTarget = _swapTarget;
+    }
+
+    /**
+     * Runs all the necessary approval functions required for a given ERC20 token.
+     * This function can be called when a new token is added to a SetToken during a
+     * rebalance.
+     *
+     * @param _token    Address of the token which needs approval
+     */
+    function approveToken(IERC20 _token, address _spender) public  isWhitelistedIssuanceModule(_spender) {
+        _safeApprove(_token, _spender, type(uint96).max);
+    }
+
+    /**
+     * Runs all the necessary approval functions required for a list of ERC20 tokens.
+     *
+     * @param _tokens    Addresses of the tokens which need approval
+     */
+    function approveTokens(IERC20[] calldata _tokens, address _spender) external {
+        for (uint256 i = 0; i < _tokens.length; i++) {
+            approveToken(_tokens[i], _spender);
+        }
+    }
+
+    /**
+     * Runs all the necessary approval functions required before issuing
+     * or redeeming a SetToken. This function need to be called only once before the first time
+     * this smart contract is used on any particular SetToken.
+     *
+     * @param _setToken    Address of the SetToken being initialized
+     */
+    function approveSetToken(ISetToken _setToken, address _issuanceModule) external {
+        address[] memory components = _setToken.getComponents();
+        for (uint256 i = 0; i < components.length; i++) {
+            // Check that the component does not have external positions
+            require(
+                _setToken.getExternalPositionModules(components[i]).length == 0,
+                "ExchangeIssuance: EXTERNAL_POSITIONS_NOT_ALLOWED"
+            );
```

Comment:
Could this require be moved to `approveToken`? Otherwise you can approve a component and bypass the external positions check.

Reply:
Like so much else, I copied this from ExchangeIssuanceV1. imo it could be moved to approveToken.
@ncitron Do you see any reason why not to move this to `approveToken` ?

Reply:
Also I would like to double check that the `approveToken` methods are supposed to be available to be called by anyone. (I assume so, since it this is also copy-paste from EI V1)

Reply:
How could we know that it is an external position in `approveToken`? It does not take a set token as a parameter

Reply:
Good point. @cgewecke I guess that answers the question, and I will just leave this as is ?

Reply:
(@ckoopmann Would it be ok to un-resolve this? Am fine with leaving things as they but would just like to flesh out all the implications and make sure I understand this....)

Reply:
So `approveToken` generically approves any token to be spent by a particular issuance module. The purpose of this is to save gas when new components are added to a set token. However, there is nothing special we can do to gatekeep what kind of tokens get approved. Even if we wanted to add a `setToken` parameter to `approveToken`, someone could just initialize a brand new set with any token they want in it as a default position (even if it is an external position for another set), and have it be approved.

Also, I'm not sure that we need to disallow tokens that are external positions. Since we now have DIM support, they should just work out of the box? I actually think it might be best to remove the check entirely. What do you think @cgewecke?

Reply:
> Since we now have DIM support, they should just work out of the box? 

@ncitron @ckoopmann Ok that sounds right to me. Agree with removal.

Reply:
I removed the external position check 👍